src/Mod/PartDesign/App/FeatureHole.cpp

```diff
@@ -1044,6 +1134,20 @@ void Hole::onChanged(const App::Property *prop)
     else if (prop == &DepthType) {
         Depth.setReadOnly((std::string(DepthType.getValueAsString()) != "Dimension"));
     }
+    else if (prop == &Depth) {
+        if (std::string(ThreadDepthType.getValueAsString()) == "Automatic") {
+            updateDiameterParam();  // make sure diameter and pitch are updated.
+            double D = threadDescription[ThreadType.getValue()][ThreadSize.getValue()].diameter;
+            ThreadDepth.setValue(std::min(5*D,  Depth.getValue() - 3 * ThreadPitch.getValue()));
+            ThreadDepth.setReadOnly(true);
+        }

```

Comment:
This is my proposal for how the thread depth can "intelligently" proposed.  In English: We need to limit the length of the thread because if the thread is too long FreeCAD will crash. Perhaps somewhat fortunate, the user will likely not need a very long threaded hole (it is not possible to manufacture anyway). We therefore limit thread to be 5 x D. The precise number of diameters can be discussed (5D is not the limit of stability). The other limitation is if it is a blind hole. In that case the thread cannot go all the way to the bottom. Instead it stops at 3 x P from the bottom (which is a good practice).

Reply:
> We therefore limit thread to be 5 x D

For example at work I use threaded M2 rods with a length of 10 cm. The holes for the rods are 2 to 3 cm deep, so >10 * D. So instead of a fixed limitation, I would rather issue a warning dialog when the depth is > 5 * D. Then the user can decide if he really needs this and accepts that he might need to wait a while until it is computed.
(and from a printing perspective, printing e.g. a 3 cm thick plate with a M3 threaded hole is doable)

> because if the thread is too long FreeCAD will crash

Can you bring that please to the developers forum? This needs to be fixed. It is acceptable that FC needs some time to calculate, but it must not crash in any circumstance, even if OCC fails.

> The other limitation is if it is a blind hole... Instead it stops at 3 x P from the bottom

Isn't there a norm that defines the "good practice"? Whenever possible we should use values or recommendations from norms.

Reply:
Note that the 5D is for the "Automatic" Thread Depth option. You will also have the possibility to specify the thread depth yourself. We could put 10D or 20D. The crashing only happens if you try to make a M1 thread that is "through all" i.e 10 m long. We can undertand how many faces such a model would have. On my system FreeCAD is killed by the kernel for using too much memory I guess.

Reply:
Regarding the norm.  I don't think there are norms for this. There might be quite official design recommendations we can find. perhaps in VDI or similar standards. Taps are certainly standardized so you can say that if you are using a standardized plug tap you will not closer than so and so from the bottom. But then you could of course 3D print your threads and then there is no limitation.

Reply:
on second thought. I had a look in my trusty Mechanical and Metal Trades Handbook. Apparently DIN 76 -1 (2004-06) gives a table. I will use that.

Reply:
![image](https://user-images.githubusercontent.com/1278189/104819647-29078d00-582f-11eb-9d4c-ec549b12f7d8.jpeg)

Reply:
Very good, so we can take the normed values instead of a hardcoded rule of thumb. So in the dialog we offer an option "automatic" that sets e1 to the ones in the DIN 76-1, if "automatic" is not checked by the user, a spinbox is enabled where he can set e1 in the range [0; hole depth].

> On my system FreeCAD is killed by the kernel for using too much memory I guess.

I am glad to hear since then there is no crash.

Reply:
Ah, we need to check that there are no differences between the DIN and the ISO. If possible we use the international standards.

Reply:
I almost agree. 

The important parameter is the thread depth, as it is also important for strength and for verifying form fit function of the bolt connection. 

If 'Automatic' is selected in the combo box (default) we will calculate:
`ThreadDepth.setValue(std::min(5*D,  Depth.getValue() - getThreadRunout(ThreadPitch.getValue())));`

Where getThreadRunout is a function that looks up e1 or e2 from the table above. I still maintain that 5*D is the max we should do for automatic. Because it makes no sense strength wise to have more. At least not for common materials. Also there will be tolerance issues for very long threaded holes. So very long threaded holes is bad practice. 

If Dimension is selected in the combo box, the a spinbox will be enabled where the user can select ThreadDepth himself between 0 and the hole depth.

Reply:
> If 'Automatic' is selected in the combo box (default) we will calculate:
> `ThreadDepth.setValue(std::min(5*D, Depth.getValue() - getThreadRunout(ThreadPitch.getValue())));`
> 
> Where getThreadRunout is a function that looks up e1 or e2 from the table above.

I don't agree because there is no reason to hardcode a 5 * D limit. "Automatic" will create a thread length of hole depth - e1 from the norm. If this result is longer than 5 * D we can issue a warning dialog that this can lead to long calculation times. If the user press OK in the warning dialog, the thread is modeled, if he declines, th thread is not modeled.

> I still maintain that 5*D is the max we should do for automatic. Because it makes no sense strength wise to have more. At least not for common materials.

This is only your use case. For many real-life applications at work, I have for example M3 x18 screws in 20 mm threaded through holes and the like. So 5 * D looks very arbitrary to me.
Also what is a common material? I print acrylic resins, others print metal parts, others ceramics.

Reply:
I admit that it is a little arbitrary, but so is having a fully threaded hole even when no standard tap can make it and still respect the thread class tolerance. We are talking about the automatic option. It is supposed to help the user get a good design. 

Even for plastics there is no structural need for more than 2.5D of thread engagement:

![image](https://user-images.githubusercontent.com/1278189/104822348-19de0a80-5842-11eb-8b92-cff7fe406674.jpeg)

For your application it will be very convenient to specify the depth manually.

Reply:
I made a proposal on how to avoid the hardcoding. I have no better proposal, I only don't like the hardcoding.

> Even for plastics there is no structural need for more than 2.5D of thread engagement

Sure there is - the costs. It is cheaper to bore a through hole through a 3 cm thick plate and subsequently thread it completely. Because otherwise someone needs to check the thread depth. This checking would be additional work increasing the costs. And in the vast majority what is cheaper will be done.

Reply:
I see, there is a different logic for these small holes. I usually work with bolts bigger than M30 and there the logic of what is costly is different. 

Anyway, I have realized another thing. When the user has specified a hole depth he has specified an intent. That should not be ignored. So I agree that we should probably remove the 5D heuristic.

Reply:
> I usually work with bolts bigger than M30

So you do heavy metal :-))

But you say you want to print in 3D. Do you operate one of these big printers? If so I will send you a private message via the forum because I am interested in this.

Reply:
Unfortunately i don't get to 3d print much at work. Apart from a brief involvement in the [LASIMM](https://www.lasimm.eu/about.html) project, it is a hobby. At work i mostly work with forgings and cast iron stuff (wind power). Pretty heavy stuff for sure ;)